### PR TITLE
OG-237 acceptanceBalance protocol

### PR DESCRIPTION
**Pull request by drortirosh**
- client should pass paymaster to relayer when calling getaddr
- getaddr's maxAcceptanceBalance should be this paymaster (or default,
if given paymaster is not trusted)
- for this the relayer caches the gasLimits of each trusted relayer..

### CODE REVIEW

File: src/common/PingResponse.ts [hunk 1]
```diff
@@ -2,11 +2,11 @@ import { Address, IntString } from '../relayclient/types/Aliases'
 
 export default interface PingResponse {
   // TODO: this should be 'worker'
```
Comment (forshtat) [inline src/common/PingResponse.ts:4]: this TODO is done

File: src/relayserver/RelayServer.ts [hunk 2]
```diff
@@ -299,12 +307,47 @@ export class RelayServer extends EventEmitter {
     process.exit(1)
   }
 
+  trustedPaymastersGasLimits: {[key: string]: PaymasterGasLimits} = {}
```
Comment (forshtat) [inline src/relayserver/RelayServer.ts:310]: Please do not define class member fields on the line 310. I know we are not in Pascal times but having methods defined in the beginning of the file is very helpful IMO.

Also, I prefer to have `Map` to typed `object`. Indexing operator is cool but somehow it does not feel right...

File: src/relayserver/RelayServer.ts [hunk 3]
```diff
@@ -299,12 +307,47 @@ export class RelayServer extends EventEmitter {
...
+   * - given that, we agree to give the requested acceptanceBudget (since breaking one of the above two "invariants"
+   *    is the only cases where the relayer will have to pay for this budget)
+   *
+   * @param paymasters list of trusted paymaster addresses
+   */
+  async _initTrustedPaymasters (paymasters: string[] = []): Promise<void> {
+    this.trustedPaymastersGasLimits = {}
+    for (const paymasterAddress of paymasters) {
+      const paymaster = await this.contractInteractor._createPaymaster(paymasterAddress)
+      const gasLimits = await paymaster.getGasLimits().catch(e => {
```
Comment (forshtat) [inline src/relayserver/RelayServer.ts:327]: You can do `.catch((e: Error) => {` and then eslint should be happy

File: src/relayserver/RelayServer.ts [hunk 4]
```diff
@@ -299,12 +307,47 @@ export class RelayServer extends EventEmitter {
...
+        // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+        throw new Error(`not a valid paymaster address in trustedPaymasters list: ${paymasterAddress}: ${e.message}`)
+      })
+      this.trustedPaymastersGasLimits[paymasterAddress.toLowerCase()] = gasLimits
+    }
+  }
+
+  _getPaymasterMaxAcceptanceBudget (paymaster?: string): IntString {
+    const limits = this.trustedPaymastersGasLimits[paymaster?.toLocaleLowerCase() ?? '']
```
Comment (forshtat) [inline src/relayserver/RelayServer.ts:336]: You don't need to check `this.trustedPaymastersGasLimits['']` IMO. Just return `maxAcceptanceBudget`

File: truffle.js [hunk 5]
```diff
@@ -3,11 +3,13 @@ require('ts-node/register/transpile-only')
 var HDWalletProvider = require('truffle-hdwallet-provider')
 var mnemonic = 'digital unknown jealous mother legal hedgehog save glory december universe spread figure custom found six'
 
-const secretMnemonicFile = './secret_mnemonic'
+// NOGIT
```
Comment (forshtat) [inline truffle.js:6]: I don't think it worked...

Comment (drortirosh) [reply truffle.js:6]: the marker was !NOGIT... I relaxed it... removing this file

File: src/relayserver/RelayServer.ts (hunk 4 above)
Comment (drortirosh) [reply src/relayserver/RelayServer.ts]: I do, since I want to return the default either if no paymaster is given, or if a wrong one is given...

**Review by forshtat — APPROVED** [review]

**Pull request merged**